Add support for .gometalinterrc

rliebz opened the pull request:
This should resolve #402 in a way that I think makes sense. Implementation details:
- The default config file is called `.gometalinterrc` and is the same format as before
- Only look for a `.gometalinterrc` if the `config` flag was not passed
- Valid locations are in the current directory, parent directories (recursively), or the home directory, in that order.
- Locations are relative to your working directory and not the file(s) being linted
- Options defined in `.gometalinterrc` have lower precedence than options passed by command line by loading them before all command-line flags

dnephin commented on main.go:
```diff
@@ -105,6 +129,51 @@ func loadConfig(app *kingpin.Application, element *kingpin.ParseElement, ctx *ki
+		prevPath, dirPath = dirPath, filepath.Dir(dirPath)
+	}
+
+	homeDir, err := getHomeDir()
```
I'm not sure about this one. If we do support a global user config it should probably be under [$XDG_CONFIG_HOME](https://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html) (at least on linux/osx). 

We could also add this later as a follow up as well, so I'm tempted to say let's only add the default project config for now.

dnephin commented on main.go:
```diff
@@ -84,8 +85,31 @@ func cliLinterOverrides(app *kingpin.Application, element *kingpin.ParseElement,
+	return loadConfigFile(*element.Value)
+}
+
+func loadConfigFile(filename string) error {
```
It would be great to move this into `config.go` now. We've been trying to split up `main.go`. Things that use `kingpin` probably still belong in `main.go`, but anything config specific could go into `config.go` (loadConfigFile, findRCFile, findRCFileInDir), and the test cases for them can move into `config_test.go`

dnephin commented on main.go:
```diff
@@ -105,6 +129,51 @@ func loadConfig(app *kingpin.Application, element *kingpin.ParseElement, ctx *ki
 	return err
 }
 
+func findRCFile() (fullPath string, found bool, err error) {
```
Naming nit  `findDefaultConfigFile()`

dnephin commented on main.go:
```diff
@@ -105,6 +129,51 @@ func loadConfig(app *kingpin.Application, element *kingpin.ParseElement, ctx *ki
+	return findRCFileInDir(homeDir)
+}
+
+func findRCFileInDir(dirPath string) (fullPath string, found bool, err error) {
```
Naming nit `findConfigFileInDir()`

rliebz commented:
@dnephin Thanks for the review. I made all the changes suggested—here's a quick summary:
- Changed to `.gometalinter.json` for the default name
- Moved and renamed functions as suggested
- Removed support for checking home directory altogether for now.

Also, since I actually already had a `.gometalinter.json` in one of my repos, I realized there should probably be a flag to disable auto-loading. I went with `--no-config` and added a few test cases, but I'd be happy to change that if you had a better approach in mind.

dnephin commented on main_test.go:
```diff
@@ -129,6 +135,61 @@ func TestPathFilter(t *testing.T) {
+	defer cleanup()
+
+	mkFile(t, tmpdir, defaultConfigPath, `{"Deadline": "3m"}`)
+	mkFile(t, tmpdir, "test-config", `{}`)
```
It would be good to set some value here so the test can check this config was loaded by checking the value.

rliebz replied:
Done.

alecthomas commented:
Nice, thanks for doing this.

alecthomas commented:
One last thing, can you add a section to the README describing this behaviour?

rliebz commented:
@alecthomas done.

alecthomas commented:
Brilliant, thanks @rliebz !
